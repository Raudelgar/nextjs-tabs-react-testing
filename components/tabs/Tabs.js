import { useState, useEffect } from 'react';
import { useRouter } from 'next/router';
import PropTypes from 'prop-types';
import { slugify } from '@/utils/index';
import styles from './Tabs.module.css';

const Tabs = ({ children, initialTab }) => {
	const [activeTab, setActiveTab] = useState(slugify(children[0].props.label));
	const router = useRouter();

	const handleClick = (e, newActiveTab) => {
		e.preventDefault();
		setActiveTab(slugify(newActiveTab));
	};
	useEffect(() => {
		if (initialTab) setActiveTab(initialTab);
	}, []);
	useEffect(() => {
		router.push(
			{
				query: { tab: slugify(activeTab) },
			},
			undefined,
			{ shallow: true }
		);
	}, [activeTab]);

	return (
		<div>
			<ul className={styles.tabs}>
				{children.map((tab) => {
					const label = tab.props.label;

					return (
						<li
							key={label}
							className={slugify(label) === activeTab ? styles.current : ''}
							// onClick={() => setActiveTab(tab.props.label)}
						>
							<a
								href='#'
								onClick={(e) => handleClick(e, label)}
								data-testid={`anchor-${slugify(label)}`}
							>
								{label}
							</a>
						</li>
					);
				})}
			</ul>
			{children.map((tabContent) => {
				const label = slugify(tabContent.props.label);
				if (label === activeTab) {
					return (
						<div
							key={label}
							data-testid={`tab-content-${label}`}
							className={styles.content}
						>
							{tabContent.props.children}
						</div>
					);
				}
			})}
		</div>
	);
};

Tabs.propTypes = {
	children: PropTypes.arrayOf(PropTypes.element).isRequired,
	initialTab: PropTypes.string,
};
Tabs.defaultProps = {
	initialTab: '',
};
export { Tabs };
