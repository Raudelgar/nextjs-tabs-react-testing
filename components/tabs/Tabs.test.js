import React from 'react';
import ReactDOM from 'react-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import { useRouter } from 'next/router';
import { Tabs } from './Tabs';

jest.mock('next/router');

describe('Tabs Component', () => {
	const component = (defaultProps) => (
		<Tabs {...defaultProps}>
			<div label='Tab 1'>
				<h2>Title Tab 1</h2>
				<p>something</p>
			</div>
			<div label='Tab 2'>
				<h2>Title Tab 2</h2>
				<p>another thing</p>
			</div>
			<div label='Tab 3'>
				<h2>Title Tab 3</h2>
				<p>hello world</p>
			</div>
		</Tabs>
	);
	beforeEach(() => {
		useRouter.mockReturnValue({
			push: jest.fn(),
		});
	});
	//Test ##################################################
	it('should match snapshot for Tabs', () => {
		const { container } = render(component({ initialTab: 'tab-3' }));
		expect(container).toMatchSnapshot();
	});
	//Test ##################################################
	it('should render the component without crash', () => {
		const { container } = render(component({ initialTab: 'tab-3' }));
		expect(container).toBeInTheDocument();
	});
	//Test ##################################################
	it('should default to the first tab content', () => {
		render(component(undefined));
		const pEl = screen.getByTestId('tab-content-tab-1').querySelector('p');
		expect(pEl.textContent).toBe('something');
	});
	//Test ##################################################
	it('should set a different tab as initial state', () => {
		render(component({ initialTab: 'tab-2' }));
		const h2El = screen.getByTestId('tab-content-tab-2').querySelector('h2');
		expect(h2El.textContent).not.toBe('Title Tab 1');
		expect(h2El.textContent).toBe('Title Tab 2');
	});
	//Test ##################################################
	it('should click a different tab and display the correct content', () => {
		render(component({ initialTab: 'tab-2' }));

		fireEvent.click(screen.getByTestId('anchor-tab-1'));
		const h2El = screen.getByTestId('tab-content-tab-1').querySelector('h2');

		expect(h2El.textContent).not.toBe('Title Tab 2');
		expect(h2El.textContent).toBe('Title Tab 1');
	});
});
