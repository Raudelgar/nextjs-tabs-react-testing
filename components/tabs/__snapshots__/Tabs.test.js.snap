// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Component should match snapshot for Tabs 1`] = `
<div>
  <div>
    <ul
      class="tabs"
    >
      <li
        class=""
      >
        <a
          data-testid="anchor-tab-1"
          href="#"
        >
          Tab 1
        </a>
      </li>
      <li
        class=""
      >
        <a
          data-testid="anchor-tab-2"
          href="#"
        >
          Tab 2
        </a>
      </li>
      <li
        class="current"
      >
        <a
          data-testid="anchor-tab-3"
          href="#"
        >
          Tab 3
        </a>
      </li>
    </ul>
    <div
      class="content"
      data-testid="tab-content-tab-3"
    >
      <h2>
        Title Tab 3
      </h2>
      <p>
        hello world
      </p>
    </div>
  </div>
</div>
`;
